File: src/imager/ImagerTypes.py
"""Типы и структуры клиентской библиотеки imager."""
from __future__ import annotations

from typing import List, TypedDict, Union

__all__ = [
    "AssetPath",
    "AssetType",
    "ImagerOptions",
    "ImagerServerOptions",
    "Segment",
    "IMAGE_FORMATS",
    "normalize_format",
    "resolve_format",
    "dedupe_formats",
]

# Форматы картинок, поддерживаемые сервисом. Всё, что не входит в этот
# список (видео и прочее), при format="auto"/"" трактуется как не-картинка.
IMAGE_FORMATS = frozenset(
    ("jpg", "jpeg", "png", "webp", "avif", "heif", "heic", "apng", "jxl", "gif")
)


def normalize_format(format: str) -> str:
    """Нормализация формата: lower-case, jpeg → jpg."""
    if format == "jpeg":
        return "jpg"
    if format.isupper():
        return format.lower()
    return format


def resolve_format(format: str, source_format: str) -> str:
    """Резолв одного формата: "auto"/"" → исходный формат, если он картинка, иначе jpg."""
    if format == "auto" or format == "":
        return source_format if source_format in IMAGE_FORMATS else "jpg"
    return normalize_format(format)


def dedupe_formats(formats: List[str]) -> List[str]:
    """Дедупликация списка форматов (jpeg → jpg, первое вхождение сохраняет позицию)."""
    seen = set()
    result = []
    for fmt in formats:
        fmt = normalize_format(fmt)
        if fmt not in seen:
            seen.add(fmt)
            result.append(fmt)
    return result


class AssetPath(TypedDict, total=False):
    """Один вариант ассета внутри paths."""

    path: str
    dpr: Union[int, float]
    width: int
    height: int


class _AssetTypeRequired(TypedDict):
    """Обязательные поля AssetType."""

    type: str
    paths: List[AssetPath]


class AssetType(_AssetTypeRequired, total=False):
    """Результат GetAsset / элемента GetAssets."""

    source_format: bool
    all_support: bool


class ImagerOptions(TypedDict, total=False):
    """Настройки клиентской части."""

    dpr: int
    format: str
    formats: List[str]
    baseURL: str
    sort: bool


class ImagerServerOptions(ImagerOptions, total=False):
    """Настройки серверной части."""

    token: str
    adminURL: str


Segment = Union[str, dict, list, tuple]


def mime_for(format: str) -> str:
    """MIME для итогового формата."""

    if format == "jpg":
        return "image/jpeg"
    return "image/" + format

File: src/imager-go/ImagerTypes.go
package imager

import "strings"

// Типы и структуры клиентской библиотеки imager (Go-реализация).
//
// Содержит:
// - структуры AssetPath / AssetType (результаты клиентских методов);
// - структуру Options (настройки конструктора).

// Один вариант ассета внутри `paths`.
//
// Все поля кроме Path опциональны: включаются в сериализацию только
// если заданы, в порядке path, dpr, width, height.
type AssetPath struct {
	Path   string  `json:"path"`
	Dpr    float64 `json:"dpr,omitempty"`
	Width  int     `json:"width,omitempty"`
	Height int     `json:"height,omitempty"`
}

// Результат GetAsset / элемента GetAssets.
//
// Опциональные поля SourceFormat/AllSupport включаются в JSON только
// при true (nil опускается через omitempty):
//
//	SourceFormat — итоговый формат совпадает с исходным форматом файла;
//	AllSupport   — формат поддерживается всеми браузерами (jpg/jpeg/gif/png).
type AssetType struct {
	Type         string      `json:"type"`
	Paths        []AssetPath `json:"paths"`
	SourceFormat *bool       `json:"source_format,omitempty"`
	AllSupport   *bool       `json:"all_support,omitempty"`
}

// boolPtr — вспомогательная функция для установки опциональных bool-полей.
func boolPtr(v bool) *bool {
	return &v
}

// imageFormats — форматы картинок, поддерживаемые сервисом. Всё, что не
// входит в этот список (видео и прочее), при format="auto"/"" трактуется
// как не-картинка.
var imageFormats = map[string]bool{
	"jpg":  true,
	"jpeg": true,
	"png":  true,
	"webp": true,
	"avif": true,
	"heif": true,
	"heic": true,
	"apng": true,
	"jxl":  true,
	"gif":  true,
}

// normalizeFormat — нормализация формата: lower-case, jpeg → jpg.
func normalizeFormat(format string) string {
	if format == "jpeg" {
		return "jpg"
	}
	return strings.ToLower(format)
}

// resolveFormat — резолв одного формата: "auto"/"" → исходный формат,
// если он картинка, иначе jpg.
func resolveFormat(format, sourceFormat string) string {
	if format == "auto" || format == "" {
		if imageFormats[sourceFormat] {
			return sourceFormat
		}
		return "jpg"
	}
	return normalizeFormat(format)
}

// dedupeFormats — дедупликация списка форматов (jpeg → jpg, первое
// вхождение сохраняет позицию).
func dedupeFormats(formats []string) []string {
	seen := make(map[string]bool, len(formats))
	result := make([]string, 0, len(formats))
	for _, fmt := range formats {
		fmt = normalizeFormat(fmt)
		if !seen[fmt] {
			seen[fmt] = true
			result = append(result, fmt)
		}
	}
	return result
}

// Настройки конструктора Imager.
type Options struct {
	Token    string
	Dpr      int
	Format   string
	Formats  []string
	BaseURL  string
	AdminURL string
}
